### DISCUSSION TITLE
Debugging: can't attach to process

### DISCUSSION THREAD

**Discussion by lukedays** [cunnane/xloil#5] [General]
Hi Steven, one more!! Thanks for being so helpful.

I'm trying to attach to the Excel process but it doesn't work. Tried in both Visual Studio & VSCode.

The following error occurs:
![image](https://user-images.githubusercontent.com/386899/186468734-b468f864-230f-433f-8064-9fb1c3d3a2ad.png)

In addition, do you plan in activating GitHub sponsors for the project?
Thanks,

**Comment by cunnane** [comment]
I tried debugging just now.  I wasn't able to replicate your error message and I did get it working, but there do seem to be some conditions:

- **Visual Studio 2019**'s attach to running process works, but the code type must be "Python code" only: with mixed mode debugging  python breakpoints aren't hit.  (Also a very strange thing happens where the number of reported threads appears to grow by 1 per second, however this doesn't seem to prevent debugging).
- **VS Code**'s attach to process works (but is laggy).  You need to add `import debugpy; debugpy.listen(...)` in your code to start a server then setup the debugging profile in VS Code to connect on the specified port.  For some reason, you need to specify an initial breakpoint with `debugpy.breakpoint()`, after that setting breakpoints in the GUI seems to work.
- **Pdb**: Putting `=xloPyDebug("pdb")` in a spreadsheet and using the `breakpoint()` function works. (However it raises a second annoying but harmless BdbQuit popup which I need to fix)
- **Pycharm** can't debug embedded python so is no help here
- **Windbg-reborn**:  I can't even get the main window to open but apparently it can debug embedded python

I'll write this up into the docs and try to play around some more to see if the situation can be improved.  I think debugging embedded python is challenging for the debugger writers which explains why it's a bit janky!

**Reply by lukedays** [reply]
Thanks for the detailed reply! I tried with VS 2022 + attach to process and VSCode + attach to server but it didn't work, I'll try some more and let you know.

**Reply by cunnane** [reply]
I'll download VS2022 and give it a crack.  Did you have the python / C++ debugging tools installed in the VS Installer workflows?

**Reply by lukedays** [reply]
I created an issue with more details! Basically attach to server works, but attach to process doesn't. No big issue though. I have everything installed, Python Tools, C++ etc

**Comment by cunnane** [comment]
I had not heard of Github sponsors before...let me look into that....
